autoscale/src/main/java/com/sequenceiq/periscope/init/InitializeMachineUserService.java

```diff
@@ -0,0 +1,48 @@
+package com.sequenceiq.periscope.init;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import javax.annotation.PostConstruct;
+import javax.inject.Inject;
+
+import org.springframework.stereotype.Component;
+
+import com.google.common.base.Strings;
+import com.sequenceiq.cloudbreak.api.endpoint.v4.stacks.response.AutoscaleStackV4Response;
+import com.sequenceiq.periscope.domain.Cluster;
+import com.sequenceiq.periscope.monitor.handler.CloudbreakCommunicator;
+import com.sequenceiq.periscope.service.AltusMachineUserService;
+import com.sequenceiq.periscope.service.ClusterService;
+
+@Component
+public class InitializeMachineUserService {
+
+    @Inject
+    private ClusterService clusterService;
+
+    @Inject
+    private AltusMachineUserService altusMachineUserService;
+
+    @Inject
+    private CloudbreakCommunicator cloudbreakCommunicator;
+
+    @PostConstruct
+    public void init() {
+        List<String> clusterCrns = clusterService.findByEnvironmentCrnOrMachineUserCrn(null, null)
+                .stream().map(Cluster::getStackCrn).collect(Collectors.toList());
+        Collection<AutoscaleStackV4Response> responses = cloudbreakCommunicator.getAutoscaleClustersByCrns(clusterCrns).getResponses();
+        responses.forEach(response -> {
+            Cluster cluster = clusterService.findOneByStackId(response.getStackId());
+            if (Strings.isNullOrEmpty(cluster.getEnvironmentCrn())) {
+                clusterService.setEnvironmentCrn(cluster.getId(), response.getEnvironmentCrn());
+                cluster = clusterService.findById(cluster.getId());
+            }
+
+            if (Strings.isNullOrEmpty(cluster.getMachineUserCrn())) {
+                altusMachineUserService.initializeMachineUserForEnvironment(cluster);
```

Comment:
I have just noticed that this is also in foreach, so we would call UMS several times and Freeipa once in a foreach iteration, so this can also cause performance issue:
pls at least check if there is bulk API for the affected UMS calls and I am pretty sure it is possible to call usersync for multiple environments

Reply:
Ack, will check for bulk UMS call and update this

Reply:
@horadla23 - is this important?
This section of the code is more of a "go populate environment / machineUser info for all clusters where it does not exist". This would happen once after Periscope is upgraded with this patch.
Remember seeing another comment about introducing a listAPI on AutoScale. Not sure if the same applies to that as well.

Reply:
I have checked count of clusters in periscope on mow-dev, it was 86, so I would guess it is not super important.
I'll let you decide if this is important to fix or not.
Anyway this is still an antipattern and can cause issues everywhere, where there is lot of clusters.
- Also it is not a big effort to use a list API for getAutoscaleClustersByCrns so please implement at least that one.
- Others can be tricky, I guess there is no batch API for UMS assignresourcerole or createmachineuser and also batch usersync call can be tricky becase different environments will have different machine user. So you can leave these as they are now.

Reply:
@horadla23  - do you mind if this moves back to the initial approach of checking these fields in ClusterStatusSyncHandler? Several reasons for this.
1) It reduces load, the need for this extra component / initialization step - which can take some time.
2) This initialization step will run on every Periscope instance, for all clusters. ClusterStatusSyncHandler is already set-up to run only for the instances assigned to a specific Periscope instance (this generally should not matter a lot, since the necessary invocations should be able to handle such parallel calls).
3) ClusterStatusSyncHandler is also taking care of scenarios where a cluster is deleted ... and as a result machine users need to be deleted.

The call in ClusterStatusSyncHandler is not expensive - since it is checking state in the Periscope database. Once a cluster is processed successfully - to add envCrn, machineUsers -  CB does not need to be invoked again for these aspects.

Reply:
sure

Reply:
updated this to use the approach discussed.